src/witness_generator/machines/block_machine.rs

```diff
@@ -0,0 +1,556 @@
+        right: &SelectedExpressions,
+        identity_index: Option<usize>,
+    ) -> EvalResult {
+        if let Some(index) = identity_index {
+            let id = &self.identities[index];
+            match id.kind {
+                IdentityKind::Polynomial => {
+                    self.process_polynomial_identity(fixed_data, id.left.selector.as_ref().unwrap())
+                }
+                IdentityKind::Plookup | IdentityKind::Permutation => {
+                    self.process_plookup(fixed_data, fixed_lookup, id)
+                }
+                _ => Err("Unsupported lookup type".to_string().into()),
+            }
+        } else {
+            self.process_outer_query(fixed_data, left, right)
+        }
+    }
+
+    /// Processes the outer query / the plookup. This function should only be called
+    /// on the acutal query row (the last one of the block).
+    fn process_outer_query(
+        &self,
+        fixed_data: &FixedData,
+        left: &[Result<AffineExpression, EvalError>],
+        right: &SelectedExpressions,
+    ) -> EvalResult {
+        assert!(self.row as usize % self.block_size == self.block_size - 1);
+        let mut errors = vec![];
+        let mut results = vec![];
+
+        // TODO how to properly hanlde the errors here?
+        // We only return them if we do not make any progress.
+
+        for (l, r) in left.iter().zip(right.expressions.iter()) {
+            match (l, self.evaluate(fixed_data, r)) {
+                (Ok(l), Ok(r)) => match (l.clone() - r).solve_with_bit_constraints(self) {
+                    Ok(result) => results.extend(result),
+                    Err(e) => errors.push(e),
+                },
+                (Err(e), Ok(_)) => errors.push(e.clone()),
+                (Ok(_), Err(e)) => errors.push(e),
+                (Err(e1), Err(e2)) => errors.extend([e1.clone(), e2]),
+            }
+        }
+        if results.is_empty() && !errors.is_empty() {
+            Err(errors.into_iter().reduce(eval_error::combine).unwrap())
+        } else {
+            Ok(results)
+        }
+    }
+
+    fn process_polynomial_identity(
+        &self,
+        fixed_data: &FixedData,
+        identity: &Expression,
+    ) -> EvalResult {
+        let evaluated = self.evaluate(fixed_data, identity)?;
+        evaluated.solve_with_bit_constraints(self).map_err(|e| {
+            let witness_data = WitnessData {
+                fixed_data,
+                data: &self.data,
+                row: self.row,
+            };
+            let formatted = evaluated.format(&witness_data);
+            if let EvalError::ConstraintUnsatisfiable(_) = e {
+                EvalError::ConstraintUnsatisfiable(format!(
+                    "Constraint is invalid ({formatted} != 0)."
+                ))
+            } else {
+                format!("Could not solve expression {formatted} = 0: {e}").into()
+            }
+        })
+    }
+
+    fn process_plookup(
```

Comment:
So these are lookups internal to this machine, which are assumed to be made against fixed columns? If so it would be good to document that